module/VuFind/src/VuFind/Auth/ShibbolethWithWAYF.php

```diff
@@ -0,0 +1,115 @@
+ *
+ * @category VuFind
+ * @package  Authentication
+ * @author   Vaclav Rosecky <vaclav.rosecky@mzk.cz>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+namespace VuFind\Auth;
+
+use VuFind\Exception\Auth as AuthException;
+
+/**
+ * Shibboleth with WAYF authentication module for authentication against
+ * multiple IdPs.
+ *
+ * @category VuFind
+ * @package  Authentication
+ * @author   Vaclav Rosecky <vaclav.rosecky@mzk.cz>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+class ShibbolethWithWAYF extends AbstractShibboleth
+{
+    /**
+     * Configured IdPs with entityId and overridden attribute mapping
+     *
+     * @var \Laminas\Config\Config
+     */
+    protected $shibbolethConfig;
+
+    /**
+     * Constructor
+     *
+     * @param \Laminas\Session\ManagerInterface $sessionManager session manager
+     * @param \Laminas\Config\Config            $shibConfig     shibboleth
+     * configuration file (shibboleth.ini)
+     */
+    public function __construct(\Laminas\Session\ManagerInterface $sessionManager,
+        \Laminas\Config\Config $shibConfig
+    ) {
+        parent::__construct($sessionManager);
+        $this->shibbolethConfig = $shibConfig;
+    }
+
+    /**
+     * Return shibboleth configuration.
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @throws \VuFind\Exception\Auth
+     * @return array shibboleth configuration
+     */
+    protected function getShibbolethConfiguration($request)
+    {
+        $entityId = $this->fetchCurrentEntityId($request);
+        if (!isset($entityId) || empty($entityId)) {
+            throw new AuthException('Missing entityId in request.');
+        }
+        $config = $this->config->Shibboleth->toArray();
+        $idpConfig = null;
+        $prefix = null;
+        foreach ($this->shibbolethConfig as $name => $configuration) {
+            if ($entityId == trim($configuration['entityId'])) {
+                $idpConfig = $configuration->toArray();
+                $prefix = $name;
+                break;
+            }
+        }
+        if ($idpConfig == null) {
+            $this->debug(
+                "Missing configuration for Idp with entityId: {$entityId})"
+            );
+            throw new AuthException('Missing configuration for IdP.');
+        }
+        foreach ($idpConfig as $key => $value) {
+            $config[$key] = $value;
+        }
+        $config['prefix'] = $prefix;
+        return $config;
+    }
+
+    /**
+     * Fetch entityId used for authentication
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object
+     *
+     * @return string entityId of IdP
+     */
+    protected function fetchCurrentEntityId($request)
```

Comment:
A minor nit: "fetch" sounds like it will access some external system to fetch the attribute. "get" would be more obvious to me.

Reply:
Agreed.